api/admin/controller/analytics_services.py

```diff
@@ -0,0 +1,122 @@
+
+    def process_get(self):
+        if flask.request.method == 'GET':
+            services = self._get_integration_info(ExternalIntegration.ANALYTICS_GOAL, self.protocols)
+            return dict(
+                analytics_services=services,
+                protocols=self.protocols,
+            )
+
+    def process_post(self):
+        name = flask.request.form.get("name")
+        protocol = flask.request.form.get("protocol")
+        fields = {"name": name, "protocol": protocol}
+        form_field_error = self.validate_form_fields(**fields)
+        if form_field_error:
+            return form_field_error
+
+        is_new = False
+        id = flask.request.form.get("id")
+        if id:
+            service = self.look_up_service_by_id(id, protocol)
+        else:
+            service, is_new = self._create_integration(
+                self.protocols, protocol, ExternalIntegration.ANALYTICS_GOAL
+            )
+
+        if isinstance(service, ProblemDetail):
+            self._db.rollback()
+            return service
+
+        name_error = self.check_name_unique(service, name, id)
+        if name_error:
+            self._db.rollback()
+            return name_error
+
+        protocol_error = self.set_protocols(service, protocol)
+        if protocol_error:
+            self._db.rollback()
+            return protocol_error
+
+        service.name = name
+
+        if is_new:
+            return Response(unicode(service.id), 201)
+        else:
+            return Response(unicode(service.id), 200)
+
+    def validate_form_fields(self, **fields):
+        """The 'name' and 'protocol' fields cannot be blank, and the protocol must
+        be selected from the list of recognized protocols."""
+
+        name = fields.get("name")
+        protocol = fields.get("protocol")
+
+        if not name:
+            return MISSING_ANALYTICS_NAME
+        if protocol and protocol not in [p.get("name") for p in self.protocols]:
+            return UNKNOWN_PROTOCOL
+
+    def look_up_service_by_id(self, id, protocol):
+        """Find an existing service, and make sure that the user is not trying to edit
+        its protocol."""
+
+        service = get_one(self._db, ExternalIntegration, id=id, goal=ExternalIntegration.ANALYTICS_GOAL)
+        if not service:
+            return MISSING_SERVICE
+        if protocol != service.protocol:
+            return CANNOT_CHANGE_PROTOCOL
+        return service
+
+    def check_name_unique(self, service, name, id):
+        """A service cannot be created with, or edited to have, the same name
+        as a service that already exists."""
+
+        service_with_name = get_one(self._db, ExternalIntegration, name=name)
+        if service_with_name and not service_with_name.id == service.id:
+            # Without checking that the IDs are different, you can't save
+            # changes to an existing service unless you've also changed its name.
+            return INTEGRATION_NAME_ALREADY_IN_USE
```

Comment:
Since you changed the behavior slightly it would be good to have a standalone test of this method. Something like this would demonstrate all the functionality, and you wouldn't have to go through calling `process_analytics_services` again.

```
def test_check_name_unique(self):

    kwargs = dict(protocol=GoogleAnalyticsProvider.__module__,
                  goal=ExternalIntegration.ANALYTICS_GOAL)
   service1, ignore = create(self._db, ExternalIntegration, name=self._str,
                             **kwargs)
   service2, ignore = create(self._db, ExternalIntegration, name=self._str,
                             **kwargs)

   # Try to change service2 so that it has the same name as service1
   # -- this is not allowed.
   result = self.controller.check_name_unique(service1, service1.name, service2.id)
   eq_(result, INTEGRATION_NAME_ALREADY_IN_USE)

   # Try to edit service1 without changing its name -- this is fine.
   eq_(
       None,
       self.controller.check_name_unique(service1, service1.name, service1.id)
   )
```

Now that I notice this, I also see that you made the same change in the metadata controller back in https://github.com/NYPL-Simplified/circulation/pull/1095/files. Add a similar test to the metadata controller and we'll be covered.